packages/store/src/patterns/encodeKey.js

```diff
@@ -140,6 +144,94 @@ const dbEntryKeyToBigint = k => {
   }
 };
 
+// `'\u0000'` is the terminator after elements.
+// `'\u0001'` is the backslash-like escape character, for
+// escaping both of these characters.
+
+const encodeArray = (array, encodeKey) => {
+  const chars = ['['];
+  for (const element of array) {
+    const enc = encodeKey(element);
+    for (const c of enc) {
+      if (c === '\u0000' || c === '\u0001') {
+        chars.push('\u0001');
+      }
+      chars.push(c);
+    }
+    chars.push('\u0000');
+  }
+  return chars.join('');
+};
+
+const decodeArray = (encodedKey, decodeKey) => {
+  assert(encodedKey.startsWith('['), X`Encoded array expected: ${encodedKey}`);
+  const elements = [];
+  const elemChars = [];
+  for (let i = 1; i < encodedKey.length; i += 1) {
+    const c = encodedKey[i];
```

Comment:
This will break codepoints into multiple code units. Why not `for (const c of encodedKey)` like encode does it?

Reply:
> This will break codepoints into multiple code units.
How does this create a problem? Can you construct an example where it does the wrong thing?

> Why not for (const c of encodedKey) like encode does it?
because I need to skip ahead one inside the loop. And it would still iterate code units anyway.

Reply:
`for (const c of encodedKey)` iterates code points not code units. It would be possible to substitute the logic of skip ahead by using a flag indicating if the next char needs to be processed as an escaped char.

That said given the simple logic of escaping here and the fact low surrogates cannot be in the range of the escaped chars, there should be no risk in this decode. It's just usually best practice to not handle code units and only deal with code points.